gateway/src/http_entrypoints.rs

```diff
@@ -0,0 +1,357 @@
+///
+/// This interface makes use of Ignition as well as the management network.
+/// SPs that are powered off (and therefore cannot respond over the
+/// management network) are represented in the output set. SPs that Ignition
+/// reports as powered on, but that do not respond within the allotted timeout
+/// will similarly be represented in the output; these will only be included in
+/// the terminal output page when the allotted timeout has expired.
+///
+/// Note that Ignition provides the full set of SPs that are plugged into the
+/// system so the gateway service knows prior to waiting for responses the
+/// expected cardinality.
+#[endpoint {
+    method = GET,
+    path = "/sp",
+}]
+async fn sp_list(
+    _rqctx: Arc<RequestContext<GatewayService>>,
+    _query: Query<TimeoutPaginationParams<SpIdentifier>>,
+) -> Result<HttpResponseOk<ResultsPage<SpInfo>>, HttpError> {
+    todo!()
+}
+
+/// Get info on an SP
+///
+/// As communication with SPs may be unreliable, consumers may specify an
+/// optional timeout to override the default.
+#[endpoint {
+    method = GET,
+    path = "/sp/{type}/{slot}",
+}]
+async fn sp_get(
+    _rqctx: Arc<RequestContext<GatewayService>>,
+    _path: Path<PathSp>,
+    _query: Query<Timeout>,
+) -> Result<HttpResponseOk<SpInfo>, HttpError> {
+    todo!()
+}
+
+/// List components of an SP
+///
+/// A component is a distinct entity under an SP's direct control. This lists
+/// all those components for a given SP.
+///
+/// As communication with SPs may be unreliable, consumers may optionally
+/// override the timeout. This interface may return a page of components prior
+/// to reaching either the timeout with the expectation that callers will keep
+/// calling this interface until the terminal page is reached. If the timeout
+/// is reached, the final call will result in an error.
+#[endpoint {
+    method = GET,
+    path = "/sp/{type}/{slot}/component",
+}]
+async fn sp_component_list(
+    _rqctx: Arc<RequestContext<GatewayService>>,
+    _path: Path<PathSp>,
+    _query: Query<TimeoutPaginationParams<PathSpComponent>>,
+) -> Result<HttpResponseOk<ResultsPage<SpComponentInfo>>, HttpError> {
+    todo!()
+}
+
+/// Get info for an SP component
+///
+/// This can be useful, for example, to poll the state of a component if
+/// another interface has changed the power state of a component or updated a
+/// component.
+///
+/// As communication with SPs maybe unreliable, consumers may specify a timeout
+/// to override the default. This interface will return an error when the
+/// timeout is reached.
+#[endpoint {
+    method = GET,
+    path = "/sp/{type}/{slot}/component/{component}",
+}]
+async fn sp_component_get(
+    _rqctx: Arc<RequestContext<GatewayService>>,
+    _path: Path<PathSpComponent>,
+    _query: Path<Timeout>,
+) -> Result<HttpResponseOk<SpComponentInfo>, HttpError> {
+    todo!()
+}
+
+// TODO: how can we make this generic enough to support any update mechanism?
```

Comment:
`&[u8]`, interpretation left as an exercise for the reader :)

(... I'm half-joking. We probably want more metadata, but that plus a version tag might be a decent start)

Reply:
Yeah, I think it's going to be some header content along with a blob... but I don't know what the former looks like.